File: binary_search.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <chrono>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <iomanip>

using namespace std;
using namespace std::chrono;

struct Row {
    int number;
    string text;
    Row(int n, const string &t) : number(n), text(t) {}
    bool operator<(const Row& other) const {
        return number < other.number;
    }
};

vector<Row> readAndSortDataset(const string &filename) {
    vector<Row> rows;
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error: Failed to open file " << filename << endl;
        return rows;
    }

    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string numPart, textPart;
        if (getline(ss, numPart, ',') && getline(ss, textPart)) {
            try {
                int number = stoi(numPart);
                rows.emplace_back(number, textPart);
            } catch (...) {
                cerr << "Warning: Skipped invalid row: " << line << endl;
            }
        }
    }
    
    sort(rows.begin(), rows.end());
    return rows;
}

int binarySearch(const vector<Row> &data, int target) {
    int left = 0;
    int right = data.size() - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (data[mid].number == target) {
            return mid;
        } else if (data[mid].number < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

void runBinarySearchTests(const vector<Row> &data) {
    int n = data.size();
    string outputFilename = "binary_search_" + to_string(n) + ".txt";
    ofstream writer(outputFilename);
    if (!writer.is_open()) {
        cerr << "Error: Failed to create output file " << outputFilename << endl;
        return;
    }

    // Prevent optimization
    volatile int dummy = 0;
    
    // Warm up
    for (int i = 0; i < 1000; i++) {
        dummy += binarySearch(data, data[0].number);
    }

    // Best case
    int bestTarget = data[n/2].number;
    auto bestStart = high_resolution_clock::now();
    for (int i = 0; i < n; i++) {
        dummy += binarySearch(data, bestTarget);
    }
    double bestTime = duration_cast<nanoseconds>(high_resolution_clock::now() - bestStart).count() / 1000000.0 / n;

    // Average case
    srand(time(0));
    auto avgStart = high_resolution_clock::now();
    for (int i = 0; i < n; i++) {
        int randomIndex = rand() % n;
        dummy += binarySearch(data, data[randomIndex].number);
    }
    double avgTime = duration_cast<nanoseconds>(high_resolution_clock::now() - avgStart).count() / 1000000.0 / n;

    // Worst case
    int worstTarget = data.back().number + 1;
    auto worstStart = high_resolution_clock::now();
    for (int i = 0; i < n; i++) {
        dummy += binarySearch(data, worstTarget);
    }
    double worstTime = duration_cast<nanoseconds>(high_resolution_clock::now() - worstStart).count() / 1000000.0 / n;

    writer << fixed << setprecision(6);
    writer << "Best case time: " << bestTime << " ms\n";
    writer << "Average case time: " << avgTime << " ms\n";
    writer << "Worst case time: " << worstTime << " ms\n";

    cout << "Results saved to " << outputFilename << endl;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <dataset_filename>" << endl;
        return 1;
    }

    string filename = argv[1];
    vector<Row> rows = readAndSortDataset(filename);
    if (rows.empty()) {
        cerr << "Error: No valid data loaded." << endl;
        return 1;
    }

    runBinarySearchTests(rows);
    return 0;
}

File: binary_search.java
import java.io.*;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class binary_search {
    
    static class DataEntry {
        final int intVal;
        final String strVal;
        
        DataEntry(int intVal, String strVal) {
            this.intVal = intVal;
            this.strVal = strVal;
        }
    }
    
    public static boolean isDatasetSorted(List<DataEntry> data) {
        for (int i = 1; i < data.size(); i++) {
            if (data.get(i).intVal < data.get(i-1).intVal) {
                return false;
            }
        }
        return true;
    }
    
    public static List<DataEntry> loadDataset(String filename) throws IOException {
        List<DataEntry> data = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length == 2) {
                    try {
                        int intVal = Integer.parseInt(parts[0].trim());
                        data.add(new DataEntry(intVal, parts[1].trim()));
                    } catch (NumberFormatException e) {
                        System.err.println("Warning: Skipped invalid row: " + line);
                    }
                }
            }
        }
        
        if (!isDatasetSorted(data)) {
            System.err.println("Error: Dataset must be pre-sorted");
            return Collections.emptyList();
        }
        
        return data;
    }
    
    public static int binarySearch(List<DataEntry> data, int target) {
        int left = 0;
        int right = data.size() - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int midVal = data.get(mid).intVal;
            
            if (midVal == target) {
                return mid;
            } else if (midVal < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return -1;
    }
    
    public static void runBinarySearchTests(List<DataEntry> data, String filename) throws IOException {
        int n = data.size();
        String outputFilename = "binary_search_" + n + ".txt";
        
        for (int i = 0; i < 1000; i++) {
            binarySearch(data, data.get(0).intVal);
        }

        // Best case (middle element)
        int bestTarget = data.get(n/2).intVal;
        long bestStart = System.nanoTime();
        for (int i = 0; i < n; i++) {
            binarySearch(data, bestTarget);
        }
        double bestTime = (System.nanoTime() - bestStart) / 1_000_000.0 / n;

        // Average case (random elements)
        Random rand = new Random();
        long avgStart = System.nanoTime();
        for (int i = 0; i < n; i++) {
            int randomIndex = rand.nextInt(n);
            binarySearch(data, data.get(randomIndex).intVal);
        }
        double avgTime = (System.nanoTime() - avgStart) / 1_000_000.0 / n;

        // Worst case (non-existent element)
        int worstTarget = data.get(n-1).intVal + 1;
        long worstStart = System.nanoTime();
        for (int i = 0; i < n; i++) {
            binarySearch(data, worstTarget);
        }
        double worstTime = (System.nanoTime() - worstStart) / 1_000_000.0 / n;

        try (PrintWriter writer = new PrintWriter(outputFilename)) {
            writer.printf("Best case time: %.6f ms%n", bestTime);
            writer.printf("Average case time: %.6f ms%n", avgTime);
            writer.printf("Worst case time: %.6f ms%n", worstTime);
        }
    }
    
    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Usage: java binary_search <dataset_filename>");
            System.exit(1);
        }
        
        try {
            List<DataEntry> data = loadDataset(args[0]);
            if (data.isEmpty()) {
                System.err.println("Error: No valid data loaded or data not sorted");
                System.exit(1);
            }
            
            runBinarySearchTests(data, args[0]);
            System.out.println("Results saved to binary_search_" + data.size() + ".txt");
            
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}